src/Diagnostics/Roslyn/Core/ApiDesign/DeclarePublicAPIAnalyzer.cs

```diff
@@ -83,132 +96,65 @@ public class DeclarePublicAPIAnalyzer : DiagnosticAnalyzer
+        public DeclarePublicAPIAnalyzer(ImmutableArray<AdditionalText> extraAdditionalFiles)
         {
-            context.RegisterCompilationStartAction(compilationContext =>
-            {
-                AdditionalText publicApiAdditionalText = TryGetPublicApiSpec(compilationContext.Options.AdditionalFiles, compilationContext.CancellationToken);
+            _extraAdditionalFiles = extraAdditionalFiles;
+        }
 
-                if (publicApiAdditionalText == null)
-                {
-                    return;
-                }
+        public DeclarePublicAPIAnalyzer()
+        {
 
-                SourceText publicApiSourceText = publicApiAdditionalText.GetText(compilationContext.CancellationToken);
-                HashSet<string> declaredPublicSymbols = ReadPublicSymbols(publicApiSourceText, compilationContext.CancellationToken);
-                HashSet<string> examinedPublicTypes = new HashSet<string>();
-                object lockObj = new object();
+        }
 
-                Dictionary<ITypeSymbol, bool> typeCanBeExtendedPubliclyMap = new Dictionary<ITypeSymbol, bool>();
-                Func<ITypeSymbol, bool> typeCanBeExtendedPublicly = type =>
-                {
-                    bool result;
-                    if (typeCanBeExtendedPubliclyMap.TryGetValue(type, out result)) return result;
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
 
-                    // a type can be extended publicly if (1) it isn't sealed, and (2) it has some constructor that is
-                    // not internal, private or protected&internal
-                    result = !type.IsSealed &&
-                        type.GetMembers(WellKnownMemberNames.InstanceConstructorName).Any(
-                            m => m.DeclaredAccessibility != Accessibility.Internal && m.DeclaredAccessibility != Accessibility.Private && m.DeclaredAccessibility != Accessibility.ProtectedAndInternal
-                        );
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var additionalFiles = compilationContext.Options.AdditionalFiles;
+            if (!_extraAdditionalFiles.IsDefaultOrEmpty)
+            {
+                additionalFiles = additionalFiles.AddRange(_extraAdditionalFiles);
+            }
 
-                    typeCanBeExtendedPubliclyMap.Add(type, result);
-                    return result;
-                };
+            ApiData shippedData;
+            ApiData unshippedData;
+            if (!TryGetApiData(additionalFiles, compilationContext.CancellationToken, out shippedData, out unshippedData))
+            {
+                return;
+            }
 
-                compilationContext.RegisterSymbolAction(symbolContext =>
+            List<Diagnostic> errors;
+            if (!ValidateApiFiles(shippedData, unshippedData, out errors))
```

Comment:
Just bail out early if ValidateApiFiles fails here, and use RegisterCompilationAction to report this error (instead of RegisterCompilationEndAction).

That way you'll get the "invalid API file" diagnostic a lot sooner in the IDE.